### PR TITLE
Implement unstable network mode (closes #2727)

### LINKED ISSUE #2727: Enhance TestCafe for working reliably in harsh network conditions
Our functional tests show that we have a problem when working with remote browsers. TestCafe should be to work reliably in environment where any request can fail unexpectedly. To overcome failed, requests we should do the following things:

* Fetch content of tested pages into the browser cache before navigating to them
* Restart failed driver requests to the server

### PR DESCRIPTION
**Pull request by AndreyBelym**

### CODE REVIEW

**Comment by AndreyBelym** [comment]
@testcafe-build-bot retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot retest

File: .travis.yml [hunk 1]
```diff
@@ -28,10 +28,10 @@ cache: yarn
 
 before_install:
   - export $(dbus-launch)
-  - curl -o- -L https://yarnpkg.com/install.sh | bash
+  - travis_retry curl -o- -L https://yarnpkg.com/install.sh | bash
```
Comment (miherlosev) [inline .travis.yml]: Does anybody measure the difference between `yarn install` and `npm install`? I guess it is a very little as compared with functional test execution time. So, we can use a standard `npm install`.

File: src/client/browser/index.js [hunk 2]
```diff
@@ -2,24 +2,43 @@
 import Promise from 'pinkie';
 import COMMAND from '../../browser/connection/command';
 import STATUS from '../../browser/connection/status';
+import { UNSTABLE_NETWORK_MODE_HEADER } from '../../browser/connection/unstable-network-mode';
 
 const HEARTBEAT_INTERVAL = 2 * 1000;
 
 var allowInitScriptExecution = false;
+var unstableNetworkMode      = false;
```
Comment (miherlosev) [inline src/client/browser/index.js]: `let unstableNetworkMode = ...`

File: src/client/driver/command-executors/execute-navigate-to.js [hunk 3]
```diff
@@ -1,19 +1,26 @@
...
+    const navigationUrl = utils.url.getNavigationUrl(command.url, window);
+
+    let ensurePagePromise = hammerhead.Promise.resolve();
+
+    if (navigationUrl && browser.isUnstableNetworkModeEnabled())
+        ensurePagePromise = browser.fetchPageToCache(navigationUrl, createNativeXHR);
+
+    return ensurePagePromise
+        .then(() => {
+            var requestBarrier = new RequestBarrier();
```
Comment (miherlosev) [inline src/client/driver/command-executors/execute-navigate-to.js]: `let requestBarrier`

File: test/functional/config.js [hunk 4]
```diff
@@ -65,11 +67,13 @@ testingEnvironments[testingEnvironmentNames.mobileBrowsers] = {
...
     },
 
+    unstableNetworkMode: true,
+
     browsers: [
         {
             realMobile: true,
             os:         'android',
-            osVersion:  '7.1',
+            osVersion:  '8.0',
```
Comment (miherlosev) [inline test/functional/config.js]: At present, the max avalible version of Android on Saucelabs is 7.1 (https://wiki.saucelabs.com/display/DOCS/Supported+Android+Emulators).
Let's update it after Andorid 8.0 will be avalible on Saucelabs.

File: src/client/browser/index.js [hunk 5]
```diff
@@ -2,24 +2,43 @@
...
+            xhr.setRequestHeader(UNSTABLE_NETWORK_MODE_HEADER, 'true');
+            xhr.setRequestHeader('accept', 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8');
+        }
+
         xhr.onreadystatechange = () => {
             if (xhr.readyState === 4) {
-                if (xhr.status === 200)
-                    resolve(xhr.responseText ? JSON.parse(xhr.responseText) : ''); //eslint-disable-line no-restricted-globals
+                if (xhr.status === 200) {
+                    let responseText = xhr.responseText ? xhr.responseText : '';
```
Comment (churkin) [inline src/client/browser/index.js]: `let responseText = xhr.responseText || '';`

File: src/client/core/utils/time-limit.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import { Promise } from '../deps/hammerhead';
+import delay from './delay';
+
+
+export default function (promise, ms) {
+    return Promise.race([promise, delay(ms).then(() => Promise.reject(new Error('Timeout')))]);
```
Comment (kirovboris) [inline src/client/core/utils/time-limit.js]: Please make a meaningful error message.

File: src/client/core/index.js [hunk 7]
```diff
@@ -46,6 +47,7 @@ exports.promiseUtils           = promiseUtils;
 exports.textSelection          = textSelection;
 exports.waitFor                = waitFor;
 exports.delay                  = delay;
+exports.timeLimit              = timeLimit;
```
Comment (kirovboris) [inline src/client/core/index.js]: I think `timeLimit` is not exactly what this function does. It's smth like `getTimeLimitedPromise`.

File: src/client/browser/index.js [hunk 8]
```diff
@@ -86,17 +105,50 @@ export function redirect (command) {
...
+
     return sendXHR(statusUrl, createXHR)
         .then(res => {
-            const redirecting = (res.cmd === COMMAND.run || res.cmd === COMMAND.idle) && !isCurrentLocation(res.url);
+            result = res;
+
+            if (result.url && isUnstableNetworkModeEnabled())
+                return fetchPageToCache(result.url, createXHR);
+
+            return Promise.resolve();
```
Comment (churkin) [inline src/client/browser/index.js]: `Promise.resolve(res); ` ?

**Comment by kirovboris** [comment]
Should we describe in the docs `unstableNetworkMode` as well?

File: src/client/driver/command-executors/execute-navigate-to.js [hunk 9]
```diff
@@ -1,19 +1,26 @@
...
-import testCafeCore from '../deps/testcafe-core';
+import { RequestBarrier, pageUnloadBarrier, browser } from '../deps/testcafe-core';
 import DriverStatus from '../status';
 
-var Promise = hammerhead.Promise;
-
-var RequestBarrier    = testCafeCore.RequestBarrier;
-var pageUnloadBarrier = testCafeCore.pageUnloadBarrier;
 
+const { createNativeXHR, utils } = hammerhead;
```
Comment (miherlosev) [inline src/client/driver/command-executors/execute-navigate-to.js]: const { createNativeXHR, utils, Promise } = hammerhead;

Comment (AndreyBelym) [reply src/client/driver/command-executors/execute-navigate-to.js]: There is bug in the hammerhead's eslint rule, it starts to show the `Use Promise from hammerhead` if you use destructuring.

Comment (miherlosev) [reply src/client/driver/command-executors/execute-navigate-to.js]: I've created issue for this case - https://github.com/LavrovArtem/eslint-plugin-hammerhead/issues/4

File: src/runner/browser-set.js [hunk 10]
```diff
@@ -132,7 +132,7 @@ export default class BrowserSet extends EventEmitter {
             BrowserSet._closeConnection(bc) :
             BrowserSet._waitIdle(bc);
 
-        var release = timeLimit(appropriateStateSwitch, this.RELEASE_TIMEOUT).then(() => remove(this.pendingReleases, release));
+        var release = getTimeLimitedPromise(appropriateStateSwitch, this.RELEASE_TIMEOUT).then(() => remove(this.pendingReleases, release));
```
Comment (miherlosev) [inline src/runner/browser-set.js]: `const release =...`

**Comment by AndreyBelym** [comment]
@testcafe-build-bot retest

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Pull request merged**